virtool/analyses/data.py

```diff
@@ -69,86 +69,106 @@ def __init__(self, db: Mongo, config, pg: AsyncEngine):
-                            "ready": True,
-                            "reference": True,
-                            "sample": True,
-                            "subtractions": True,
-                            "updated_at": True,
-                            "user": True,
-                        },
-                        "total_count": {
-                            "$arrayElemAt": ["$total_count.total_count", 0]
-                        },
-                        "found_count": {
-                            "$arrayElemAt": ["$found_count.found_count", 0]
-                        },
-                    }
-                },
-            ]
-        ):
+        pipeline = [
+            {
+                "$facet": {
+                    "total_count": [{"$count": "total_count"}],
+                    "found_count": [{"$count": "found_count"}],
+                    "data": [
+                        {"$sort": sort},
+                        {"$skip": skip_count},
+                        {"$limit": per_page},
+                        *lookup_nested_subtractions(),
+                        *lookup_nested_reference_by_id(),
+                    ],
+                }
+            },
+            {
+                "$project": {
+                    "data": {
+                        "_id": True,
+                        "workflow": True,
+                        "created_at": True,
+                        "index": True,
+                        "job": True,
+                        "ready": True,
+                        "reference": True,
+                        "sample": True,
+                        "subtractions": True,
+                        "updated_at": True,
+                        "user": True,
+                    },
+                    "total_count": {"$arrayElemAt": ["$total_count.total_count", 0]},
+                    "found_count": {"$arrayElemAt": ["$found_count.found_count", 0]},
+                }
+            },
+        ]
+
+        async for paginate_dict in self._mongo.analyses.aggregate(pipeline):
             data = paginate_dict["data"]
             found_count: int = paginate_dict.get("found_count", 0)
             total_count: int = paginate_dict.get("total_count", 0)
 
-        per_document_can_read = await asyncio.gather(
-            *[
-                virtool.samples.db.check_rights(
-                    self._mongo, document["sample"]["id"], client, write=False
+        if sample_id is not None:
+            can_read = [
+                virtool.samples.db.check_rights_error_check(
+                    self._mongo,
+                    sample_id,
+                    client,
+                    write=False,
+                )
+            ]
+            found_count = sum(
+                document["sample"]["id"] == sample_id for document in data
+            )
+        else:
+            can_read = [
+                virtool.samples.db.check_rights_error_check(
```

Comment:
A permission error raised here is a problem for clients who haven't asked for a specific sample_id.